scheduler/system_util.go

```diff
@@ -291,6 +287,64 @@ func diffSystemAllocsForNode(
 	return result
 }
 
+// ensureSingleSystemAlloc enforces the invariant that a system or sysbatch job
+// should never have more than one allocation in a desired-running state.
+func ensureSingleSystemAlloc(result *diffResult) {
+
+	if len(result.ignore)+len(result.update)+len(result.reconnecting) <= 1 {
+		return
+	}
+
+	// sort by JobModifyIndex, then CreateIndex
+	sortTuples := func(tuples []allocTuple) {
+		sort.Slice(tuples, func(i, j int) bool {
+			if tuples[i].Alloc.Job.JobModifyIndex == tuples[j].Alloc.Job.JobModifyIndex {
+				return tuples[i].Alloc.CreateIndex > tuples[j].Alloc.CreateIndex
+			}
+			return tuples[i].Alloc.Job.JobModifyIndex > tuples[j].Alloc.Job.JobModifyIndex
```

Comment:
golf: sometimes is is nice to assign intermediate variables in these so the statements are less verbose, e.g.

```
func(i, j int) bool {
  I, J := tuples[i].Alloc, tuples[j].Alloc
  if I.Job.ModifyIndex == J.Job.ModifyIndex {
  ...
}
```

Reply:
actually we probably don't need to sort these - we only need to iterate through and find the index of the max, then iterate again copying to []stop but skipping the max, then creating the [1] with the max in it

if these slices are always small the Sort is probably fine and definitely more readable

Reply:
The inputs are the ignore/update/reconnecting allocations that were assigned to a single node, so that constrains the size of these slices quite a bit. Most of the time if we don't hit the bail-out early condition I'd expect there to be only 2 or 3 items between all 3 slices.